pyscriptjs/src/plugins/pyterminal.ts

```diff
@@ -70,59 +83,182 @@ export class PyTerminalPlugin extends Plugin {
-            } else {
-                this.autoShowOnNextLine = false;
-            }
+            this.setupPosition(app);
+        }
 
+        // implementation of the Stdio interface
+        stdout_writeline(msg: string) {
+            this.outElem.innerText += msg + '\n';
             if (this.isDocked()) {
-                this.classList.add('py-terminal-docked');
+                this.scrollTop = this.scrollHeight;
+            }
+            if (this.autoShowOnNextLine) {
+                this.classList.remove('py-terminal-hidden');
+                this.autoShowOnNextLine = false;
             }
+        }
 
-            logger.info('Registering stdio listener');
-            app.registerStdioListener(this);
+        stderr_writeline(msg: string) {
+            this.stdout_writeline(msg);
         }
+        // end of the Stdio interface
+    }
+
+    return PyTerminalPre;
+}
+
+declare const Terminal: typeof TerminalType;
+
+function make_PyTerminal_xterm(app: PyScriptApp) {
+    /** The <py-terminal> custom element, which automatically register a stdio
+     *  listener to capture and display stdout/stderr
+     */
+    class PyTerminalXterm extends PyTerminalBaseClass {
+        outElem: HTMLDivElement;
+        _moduleResolved: boolean;
+        xtermReadyPromise: Promise<TerminalType>;
+        xterm: TerminalType;
+        cachedStdOut: Array<string>;
+        cachedStdErr: Array<string>;
+        _xterm_cdn_base_url = 'https://cdn.jsdelivr.net/npm/xterm@5.1.0';
 
-        isAuto() {
-            return this.hasAttribute('auto');
+        constructor() {
+            super();
+            this.cachedStdOut = [];
+            this.cachedStdErr = [];
+
+            // While this is false, store writes to stdout/stderr to a buffer
+            // when the xterm.js is actually ready, we will "replay" those writes
+            // and set this to true
+            this._moduleResolved = false;
+
+            //Required to make xterm appear properly
+            this.style.width = '100%';
+            this.style.height = '100%';
         }
 
-        isDocked() {
-            return this.hasAttribute('docked');
+        async connectedCallback() {
+            //guard against initializing a tag twice
+            if (knownPyTerminalTags.has(this)) return;
+
+            this.outElem = document.createElement('div');
+            //this.outElem.className = 'py-terminal';
+            this.appendChild(this.outElem);
+
+            this.setupPosition(app);
+
+            this.xtermReadyPromise = this._setupXterm();
+            await this.xtermReadyPromise;
+            knownPyTerminalTags.add(this);
+        }
+
+        /**
+         * Fetch the xtermjs library from CDN an initialize it.
+         * @private
+         * @returns the associated xterm.js Terminal
+         */
+        async _setupXterm() {
+            // eslint-disable-next-line
+            // @ts-ignore
+            if (globalThis.Terminal == undefined) {
```

Comment:
nit: this looks like a guard that works if we're the only script on the page that could import `xterm` ... as the Web is very wild, shouldn't this be an `init` or similar flag private to this module scope? trusting globals in JS has never been a great idea, so I wonder if it makes sense to stop doing that for new code we land in prod.

Reply:
Since xterm is loading in a way it's not strictly designed to (i.e. it's meant to be used as a node module and not lazy loaded), I'm not sure how to load it in a way that doesn't end up dumping Teminal into the globals. That said, you did make me realize that we should be separating the check for whether a specific py-terminal has been initialized from whether there's a Terminal global object, that's included in the latest commit.